pkg/sql/ttl/ttljob/ttljob_processor.go

```diff
@@ -0,0 +1,341 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package ttljob
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
+	"github.com/cockroachdb/cockroach/pkg/util/quotapool"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
+)
+
+func (t rowLevelTTLResumer) work(
+	ctx context.Context,
+	db *kv.DB,
+	rangeConcurrency int64,
+	execCfg *sql.ExecutorConfig,
+	details jobspb.RowLevelTTLDetails,
+	descsCol *descs.Collection,
+	knobs sql.TTLTestingKnobs,
+	initialVersion descpb.DescriptorVersion,
+	pkColumns []string,
+	relationName string,
+	selectBatchSize int64,
+	deleteBatchSize int64,
+	deleteRateLimit int64,
+	aost *tree.DTimestampTZ,
+	ttlExpression string,
+) (int64, error) {
+
+	deleteRateLimiter := quotapool.NewRateLimiter(
+		"ttl-delete",
+		quotapool.Limit(deleteRateLimit),
+		deleteRateLimit,
+	)
+
+	rowCount := int64(0)
+
+	var pkTypes []*types.T
+	var rangeSpan, entirePKSpan roachpb.Span
+	var labelMetrics bool
+	if err := db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		desc, err := descsCol.GetImmutableTableByID(
+			ctx,
+			txn,
+			details.TableID,
+			tree.ObjectLookupFlagsWithRequired(),
+		)
+		if err != nil {
+			return err
+		}
+
+		primaryIndexDesc := desc.GetPrimaryIndex().IndexDesc()
+		pkColumns = primaryIndexDesc.KeyColumnNames
+		for _, id := range primaryIndexDesc.KeyColumnIDs {
+			col, err := desc.FindColumnWithID(id)
+			if err != nil {
+				return err
+			}
+			pkTypes = append(pkTypes, col.GetType())
+		}
+
+		if !desc.HasRowLevelTTL() {
+			return errors.Newf("unable to find TTL on table %s", desc.GetName())
+		}
+
+		rowLevelTTL := desc.GetRowLevelTTL()
+		labelMetrics = rowLevelTTL.LabelMetrics
+
+		tn, err := descs.GetTableNameByDesc(ctx, txn, descsCol, desc)
+		if err != nil {
+			return errors.Wrapf(err, "error fetching table relation name for TTL")
+		}
+
+		relationName = tn.FQString()
+		entirePKSpan = desc.PrimaryIndexSpan(execCfg.Codec)
+		rangeSpan = entirePKSpan
+		return nil
+	}); err != nil {
+		return rowCount, err
+	}
+
+	metrics := execCfg.JobRegistry.MetricsStruct().RowLevelTTL.(*RowLevelTTLAggMetrics).loadMetrics(
+		labelMetrics,
+		relationName,
+	)
+
+	g := ctxgroup.WithContext(ctx)
```

Comment:
`g` and `ch` cannot be shared with the ttljob because `work(...)` will run on different machines.